File: ios/CryptoSavingsTracker/Models/Transaction.swift
//
//  Transaction.swift
//  CryptoSavingsTracker
//
//  Created by Assistant on 11/08/2025.
//

import SwiftData
import Foundation

@Model
final class Transaction {
    init(
        amount: Double,
        asset: Asset,
        date: Date = Date(),
        source: TransactionSource = .manual,
        externalId: String? = nil,
        counterparty: String? = nil,
        comment: String? = nil
    ) {
        self.id = UUID()
        self.amount = amount
        self.date = date
        self.asset = asset
        self.comment = comment
        self.sourceRawValue = source.rawValue
        self.externalId = externalId
        self.counterparty = counterparty
    }

    var id: UUID = UUID()
    var amount: Double = 0.0
    var date: Date = Date()
    var sourceRawValue: String = TransactionSource.manual.rawValue
    var externalId: String?
    var counterparty: String?
    var comment: String?
    
    var asset: Asset?

    var source: TransactionSource {
        TransactionSource(rawValue: sourceRawValue) ?? .manual
    }
}

File: android/app/src/main/java/com/xax/CryptoSavingsTracker/domain/model/Transaction.kt
package com.xax.CryptoSavingsTracker.domain.model

import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter

/**
 * Domain model representing a transaction (deposit or withdrawal).
 */
data class Transaction(
    val id: String,
    val assetId: String,
    val amount: Double,
    val dateMillis: Long,
    val source: TransactionSource,
    val externalId: String?,
    val counterparty: String?,
    val comment: String?,
    val createdAt: Long
) {
    /**
     * Whether this is a deposit (positive amount) or withdrawal (negative amount)
     */
    val isDeposit: Boolean
        get() = amount >= 0

    /**
     * Absolute value of the amount
     */
    val absoluteAmount: Double
        get() = kotlin.math.abs(amount)

    /**
     * Get the date as LocalDateTime
     */
    fun dateTime(zone: ZoneId = ZoneId.systemDefault()): LocalDateTime {
        return Instant.ofEpochMilli(dateMillis).atZone(zone).toLocalDateTime()
    }

    /**
     * Format the date for display
     */
    fun formattedDate(pattern: String = "MMM d, yyyy", zone: ZoneId = ZoneId.systemDefault()): String {
        return dateTime(zone).format(DateTimeFormatter.ofPattern(pattern))
    }

    /**
     * Format the date and time for display
     */
    fun formattedDateTime(
        pattern: String = "MMM d, yyyy 'at' h:mm a",
        zone: ZoneId = ZoneId.systemDefault()
    ): String {
        return dateTime(zone).format(DateTimeFormatter.ofPattern(pattern))
    }
}

/**
 * Source of the transaction.
 * iOS has: manual, onChain
 * Android extends with: import (for CSV import feature)
 */
enum class TransactionSource(val rawValue: String) {
    MANUAL("manual"),
    ON_CHAIN("onChain"),
    IMPORT("import");  // Android extension - not in iOS

    companion object {
        fun fromString(value: String): TransactionSource {
            return entries.find {
                it.rawValue.equals(value, ignoreCase = true) ||
                it.name.equals(value, ignoreCase = true)
            } ?: MANUAL
        }
    }

    fun displayName(): String = when (this) {
        MANUAL -> "Manual Entry"
        ON_CHAIN -> "On-Chain"
        IMPORT -> "Imported"
    }
}
